jupyter_ydoc/ystdin.py

```diff
@@ -0,0 +1,36 @@
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+from pycrdt import Map, Text
+
+
+def add_stdin(cell: Map, prompt: str = "", password: bool = False) -> None:
+    """
+    Adds an stdin Map in the cell outputs.
+
+    Schema:
+
+    .. code-block:: json
+
+        {
+            "state": Map[
```

Comment:
This is a question we do not need to answer now: should there be an owner of the input box too? So that the input can be provided only by the user who executed the code.

Reply:
If we think inputs are collaborative, then anybody should be able to fill them, right?
Otherwise I don't think this is really compatible with CRDTs, that are essentially shared structures.

Reply:
A teacher may want students to see what they are typing, but not allow students to type in into their box.

Reply:
At first sight it looks like the same issue as with cell code: there is no way to prevent a student from changing the code. Why would it be different with a input? If it's because it's the teacher who executed the cell, well this information is not present in the shared model, and I don't think it should be. And anyway, a student can re-execute the cell, and now should they be able to type in into the box? It doesn't really make sense.
A clearer model would be that the whole notebook is read-only for students. In this case we could ignore their Y-updates so that they can't change the shared model. This would make use of user permissions.
This is a general constraint of CRDTs, everybody sees the same data, unlike in a server/client architecture where a client makes a request and receives a dedicated reply. For instance, paginated results are possible with HTTP but not with CRDTs.